pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -0,0 +1,88 @@
+package oauthclient
+
+import (
+	"strconv"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+
+	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	oauthapi "github.com/openshift/origin/pkg/oauth/api"
+	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+)
+
+const UseAsOAuthClientSecretAnnotation = "serviceaccounts.openshift.io/use-as-client-secret"
+
+type saOAuthClientAdapter struct {
+	saClient     kclient.ServiceAccountsNamespacer
+	secretClient kclient.SecretsNamespacer
+
+	delegate oauthclient.Getter
+}
+
+var _ oauthclient.Getter = &saOAuthClientAdapter{}
+
+func NewServiceAccountOauthClientGetter(saClient kclient.ServiceAccountsNamespacer, secretClient kclient.SecretsNamespacer, delegate oauthclient.Getter) oauthclient.Getter {
+	return &saOAuthClientAdapter{saClient: saClient, secretClient: secretClient, delegate: delegate}
+}
+
+func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oauthapi.OAuthClient, error) {
+	saNamespace, saName, err := serviceaccount.SplitUsername(name)
+	if err != nil {
+		return a.delegate.GetClient(ctx, name)
+	}
+
+	tokens, err := a.getServiceAccountTokens(saNamespace, saName)
+	if err != nil {
+		return nil, err
+	}
+
+	saClient := &oauthapi.OAuthClient{
+		ObjectMeta:        kapi.ObjectMeta{Name: name},
+		ScopeRestrictions: getScopeRestrictionsFor(saNamespace, saName),
+		AdditionalSecrets: tokens,
+
+		// TODO should we allow redirecting to localhost
+		// BEFORE MERGE: update this to allow https redirection to any
+		// 1. service IP (useless in general)
+		// 2. service DNS (useless in general)
+		// 3. route DNS (useful)
```

Comment:
@smarterclayton I seem to recall that `route.status` isn't a reliable place to go to find the actual name of the routes that I have.  I'm not sure where else to get it inside the API, but I need to have that information so that these flows will actually work.

I might want to let a user further restrict the list (probably not though), but I don't want to trust them to enter correct information here.  This will be much easier to use and less error-prone if we auto-create the list for them.